crates/ruff/src/rules/ruff/rules/invalid_index_type.rs

```diff
@@ -0,0 +1,212 @@
+use rustpython_parser::ast::{Constant, Expr, ExprConstant, ExprSlice, ExprSubscript, Ranged};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use std::fmt;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for indexed access to lists, strings, tuples, bytes, and comprehensions
+/// using a type other than an integer or slice.
+///
+/// ## Why is this bad?
+/// Only integers or slices can be used as indices to these types. Using
+/// other types will result in a `TypeError` at runtime and a `SyntaxWarning` at
+/// import time.
+///
+/// ## Example
+/// ```python
+/// var = [1, 2, 3]["x"]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// var = [1, 2, 3][0]
+/// ```
+#[violation]
+pub struct InvalidIndexType {
+    value_type: String,
+    index_type: String,
+    is_slice: bool,
+}
+
+impl Violation for InvalidIndexType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let InvalidIndexType {
+            value_type,
+            index_type,
+            ..
+        } = self;
+        if self.is_slice {
+            format!("Slice in indexed access to type `{value_type}` uses type `{index_type}` instead of an integer.")
+        } else {
+            format!(
+                "Indexed access to type `{value_type}` uses type `{index_type}` instead of an integer or slice."
+            )
+        }
+    }
+}
+
+/// RUF015
+pub(crate) fn invalid_index_type(checker: &mut Checker, expr: &ExprSubscript) {
+    let ExprSubscript {
+        value,
+        slice: index,
+        ..
+    } = expr;
+
+    // Check the value being indexed is a list, tuple, string, f-string, bytes, or comprehension
+    if !matches!(
+        value.as_ref(),
+        Expr::List(_)
+            | Expr::ListComp(_)
+            | Expr::Tuple(_)
+            | Expr::JoinedStr(_)
+            | Expr::Constant(ExprConstant {
+                value: Constant::Str(_) | Constant::Bytes(_),
+                ..
+            })
+    ) {
+        return;
+    }
+
+    // The value types supported by this rule should always be checkable
+    let value_type = CheckableExprType::try_from(value)
+        .expect("Expected indexed expression to be a checkable type.");
```

Comment:
To me it would be a bit more natural to do this before the `if !matches` (like `let Some(value_type) = ... else { return; };`), then have an `if !matches(value_type, CheckableExprType::List | ...)` to enforce the further narrowing. That way the `expect` is unnecessary as the condition is impossible. But it's not blocking.

Reply:
I did this before then moved it back to an `expect`, but it's confuses the logic of the rule. Like...

If the value is a list, byte, string, or tuple then we want to enforce this rule.

To display a violation, we happen to use `CheckableExprType` to generate a type name for the value. However, this is just for display purposes not a limitation of the rule itself. 

If someone were to extend the rule to apply to a new type, e.g. byte arrays, I would not expect this function to exit early because there is not an implementation for displaying the type name. In this case, the `expect` would point a developer to the next step in their implementation.

🤷‍♀️ it does feel like a subtle stylistic choice in the end — maybe you shouldn't have sent me that BurntSushi article 😝

I'm kind of curious to see what lessons I learn from making my own stylistic choices but I'm happy to follow precedent for the project too!

Reply:
This makes sense. I'd say run with what you have here! I tend to be pretty defensive around `expect` and `unwrap` (which are ~equivalent) because they're not recoverable in any way, so e.g. if we shipped a release that had an oversight here, Ruff would bail entirely on files that hit this codepath.

The "right" answer here may be to extend `CheckableExprType` to instead implement all expression types (so it doesn't return `Option`), or perhaps to use a debug assertion so that we see this failure in debug builds but fail silently in releases. But it's also ok for now, we know that it's a superset of expressions on the `match` anyway. (And we're obviously overdoing it on this one point just for educational / discussion purposes, it's not a sticking point in practice :))

Reply:
> or perhaps to use a debug assertion so that we see this failure in debug builds but fail silently in releases.

I'm pretty happy with that as an approach.

Although... I'd also _want_ a user to report a failure here because it's a bug in Ruff. I guess it's a matter of if we want to panic and make the rule entirely unusable for their code or fail silently and _hope_ they notice that the rule isn't raising a violation. I can understand preferring the second as a safer user experience :)

Reply:
I wanted `debug_panic!(...)` but I guess `debug_assert!(false, ...)` will do https://github.com/astral-sh/ruff/pull/5602/commits/e058e240574053c900129bcfae2c18016ab61563

Reply:
I'm happy with whatever! If it were me on a desert island, I _guess_ I'd write something like `let Some(value_type) = CheckableExprType::try_from(value) else { return; };` but I'm not convinced that's the best option. This looks good to me.

(Again just to complete the picture, I _think_ if we hit a panic here, the behavior the user would see is that we'd throw up a message asking them to file an issue and wouldn't print any violations for the failing file, but would print out violations for all other files as usual.)